### PR TITLE
Fix averages for mix of facilitator and non facilitator surveys

### PR DESCRIPTION
**Pull request by mehalshah**
We were underreporting survey results for some teachers. This is because some surveys had facilitator specific answers
```
question
  facilitator_name
    answer
```
and some did not 
```
question
  answer
```

When there was a mix of these responses, we would only take the sum of the non-facilitator specific answers, but divide it by the total number of responses leading us to underreport results.

The short term solution is to do the following
1) Determine if we are in a situation where there is a mix of facilitator-specific and non-facilitator specific answers
2) Update the map of facilitator response counts accordingly
3) When computing response sums, when we have a non-facilitator specific answer for a facilitator-specific question, assume that the answer applies to all facilitators

### CODE REVIEW

File: dashboard/lib/api/v1/pd/workshop_score_summarizer.rb [hunk 1]
```diff
@@ -151,15 +151,42 @@ def get_score_for_workshops(workshops:, include_free_responses:, facilitator_nam
   # Return a hash of facilitator name to how many responses they have
   # @param responses [Array(Hash)] List of responses in hash form
   # @return [Hash] Hash of names to response counts
-  def calculate_facilitator_name_frequencies(responses)
-    if responses.first['how_often_given_feedback_s'].is_a? Hash
+  def calculate_facilitator_name_frequencies(responses, workshops)
```
Comment (aoby) [inline dashboard/lib/api/v1/pd/workshop_score_summarizer.rb]: add a `@param` for `workshops`

File: dashboard/lib/api/v1/pd/workshop_score_summarizer.rb [hunk 2]
```diff
@@ -151,15 +151,42 @@ def get_score_for_workshops(workshops:, include_free_responses:, facilitator_nam
...
+      # facilitator specific responses. So we assume that each response is for all
+      # facilitators and add to responses accordingly
+
+      # We need a map of workshop_ids to facilitators. Here, we use it to update the
+      # response count for non-facilitator specific responses
+      workshops_to_facilitators = workshops.joins(:facilitators).includes(:facilitators).map {|x| [x.id, x.facilitators.map(&:name)]}.to_h
+
+      facilitator_names = (facilitator_names + workshops_to_facilitators.values.flatten).uniq
+
+      facilitator_name_responses = Hash[facilitator_names.compact.map {|facilitator| [facilitator, 0]}]
```
Comment (aoby) [inline dashboard/lib/api/v1/pd/workshop_score_summarizer.rb]: nit: be consistent with line 167, either both `map(...).to_h` or `Hash[map(...)]`. I prefer the former, `.to_h` for readability.

File: dashboard/lib/api/v1/pd/workshop_score_summarizer.rb [hunk 3]
```diff
@@ -171,23 +198,25 @@ def initialize_response_summaries(facilitators, facilitator_name_filter = nil)
...
-        Hash[FACILITATOR_SPECIFIC_MULTIPLE_CHOICE_QUESTIONS.map {|question| [question, Hash[facilitators.map {|facilitator| [facilitator, 0]}]]}]
-      end
+    if facilitators
+      # Initialize a hash of facilitator specific questions to hashes of facilitator->answer sums.
+      facilitator_specific_response_sums =
+        if facilitator_name_filter
+          Hash[FACILITATOR_SPECIFIC_MULTIPLE_CHOICE_QUESTIONS.map {|question| [question, {facilitator_name_filter => 0}]}]
+        else
+          Hash[FACILITATOR_SPECIFIC_MULTIPLE_CHOICE_QUESTIONS.map {|question| [question, Hash[facilitators.map {|facilitator| [facilitator, 0]}]]}]
+        end
```
Comment (aoby) [inline dashboard/lib/api/v1/pd/workshop_score_summarizer.rb:217]: These nested hashes are hard to parse on one line. I suggest blowing this up a bit. 

Also nit: again I tend to prefer `.to_h` to `Hash[` but that's your call.

How about something more like this:
```ruby
FACILITATOR_SPECIFIC_MULTIPLE_CHOICE_QUESTIONS.map do |question| 
  [
    question, 
    facilitators.map {|facilitator| [facilitator, 0]}.to_h
  ]
end.to_h
```

or maybe
```ruby
scores_per_facilitator = facilitators.map {|facilitator| [facilitator, 0]}.to_h
FACILITATOR_SPECIFIC_MULTIPLE_CHOICE_QUESTIONS.map {|q| [q, scores_per_facilitator.dup]}.to_h
```

File: dashboard/lib/api/v1/pd/workshop_score_summarizer.rb [hunk 4]
```diff
@@ -171,23 +198,25 @@ def initialize_response_summaries(facilitators, facilitator_name_filter = nil)
...
 
-    free_response_summary = Hash[FREE_RESPONSE_QUESTIONS.map {|question| [question, []]}]
+      # Initialize a hash of free response questions to hashes of facilitators->answers
+      facilitator_specific_free_response_sums =
+        if facilitator_name_filter
+          Hash[FACILITATOR_SPECIFIC_FREE_RESPONSE_QUESTIONS.map {|question| [question, {facilitator_name_filter => []}]}]
+        else
+          Hash[FACILITATOR_SPECIFIC_FREE_RESPONSE_QUESTIONS.map {|question| [question, Hash[facilitators.map {|facilitator| [facilitator, []]}]]}]
+        end
```
Comment (aoby) [inline dashboard/lib/api/v1/pd/workshop_score_summarizer.rb:225]: Please split up these long lines. They are over our [max 120 chars](https://github.com/code-dot-org/code-dot-org/blob/staging/STYLEGUIDE.md#all-languages), and difficult to read. And like above, the nested hashes are particularly difficult to parse in this form.

File: dashboard/lib/api/v1/pd/workshop_score_summarizer.rb [hunk 5]
```diff
@@ -151,15 +151,42 @@ def get_score_for_workshops(workshops:, include_free_responses:, facilitator_nam
...
+          response['who_facilitated_ss'].each do |facilitator|
+            facilitator_name_responses[facilitator] += 1
+          end
+        else
+          workshops_to_facilitators[response['workshop_id_i']].each {|facilitator| facilitator_name_responses[facilitator] += 1}
+        end
+      end
+
+      return facilitator_name_responses, workshops_to_facilitators
+    else
```
Comment (aoby) [inline dashboard/lib/api/v1/pd/workshop_score_summarizer.rb:193]: This seems unnecessarily complex to me. Why do we need the special cases? The above `elsif` block handles a mix of facilitator-specific and non-facilitator-specific questions, so can't that be the general (and only) case? Why do we need the `else` below? 

It also looks like we don't need the initial `if facilitator_names == [nil]` case, again this could be handled generally in one branch, though it could also be a useful short circuit.

File: dashboard/lib/api/v1/pd/workshop_score_summarizer.rb [hunk 6]
```diff
@@ -150,16 +150,52 @@ def get_score_for_workshops(workshops:, include_free_responses:, facilitator_nam
 
   # Return a hash of facilitator name to how many responses they have
   # @param responses [Array(Hash)] List of responses in hash form
+  # @param workshops [Array(Pd::Workshop)] List of workshops
   # @return [Hash] Hash of names to response counts
-  def calculate_facilitator_name_frequencies(responses)
-    if responses.first['how_often_given_feedback_s'].is_a? Hash
+  # @return [Hash] Hash of workshops_to_facilitators - returned when there is a mix of
```
Comment (aoby) [inline dashboard/lib/api/v1/pd/workshop_score_summarizer.rb:155]: 2 returns?

**Review by aoby — COMMENTED** [review]
Overall, this looks ok but some parts are difficult to read or reason about and I have a few suggestions on clarity.

Zooming out, instead of special-casing the old format in 2 locations, consider first coercing the response data to the new format then processing everything simply and generically. In other words, do a 2-pass process: first expand any non-facilitator-specific answers to each facilitator, then process the enhanced results in the normal way.

Thoughts?

**Comment by mehalshah** [comment]
>consider first coercing the response data to the new format then processing everything simply and generically.

I hear you. Worth a try, I'll get back to you on how well that works in practice
